frontend/src/composables/use-floating-ui.ts

```diff
@@ -0,0 +1,115 @@
+import { ref, watch, Ref, ToRefs } from "vue"
+import {
+  autoUpdate,
+  computePosition,
+  detectOverflow,
+  flip,
+  limitShift,
+  MiddlewareState,
+  offset,
+  Placement,
+  shift,
+  Strategy,
+} from "@floating-ui/dom"
+
+export type PopoverContentProps = {
+  visible: boolean
+  hide: () => void
+  hideOnEsc: boolean
+  hideOnClickOutside: boolean
+  autoFocusOnShow: boolean
+  autoFocusOnHide: boolean
+  triggerElement: HTMLElement | null
+  placement: Placement
+  strategy: Strategy
+  clippable: boolean
+  trapFocus: boolean
+  zIndex: number | string
+}
+
+type Props = {
+  floatingElRef: Ref<HTMLElement | null>
+  floatingPropsRefs: ToRefs<PopoverContentProps>
+}
+
+// A constant offset to ensure there's a gap between
+// the floating element and the window border.
+const FloatingOffset = 20
+
+export function useFloatingUi({ floatingElRef, floatingPropsRefs }: Props) {
+  const maxHeightRef = ref<number | null>(null)
+  const style = ref<Record<string, string> | null>(null)
+  const cleanup = ref<null | (() => void)>(null)
+
+  const buildAutoUpdate = (
+    triggerElement: HTMLElement,
+    floatingElement: HTMLElement
+  ) => {
+    return autoUpdate(triggerElement, floatingElement, () => {
+      computePosition(triggerElement, floatingElement, {
+        placement: floatingPropsRefs.placement.value,
+        middleware: [
+          offset(8),
+          flip(),
+          shift({ limiter: limitShift() }),
+          detectMaxHeight,
+        ],
+      }).then(({ x, y }) => {
+        style.value = {
+          left: `${x}px`, // floating-ui is handling RTL, so `left` is set correctly here.
+          top: `${y}px`,
+        }
+      })
+    })
+  }
+
+  watch(
+    [
+      floatingElRef,
+      floatingPropsRefs.triggerElement,
+      floatingPropsRefs.visible,
+    ] as const,
+    ([floatingElement, triggerElement, visible], _, onInvalidate) => {
+      if (!triggerElement || !floatingElement) {
+        cleanup.value?.()
```

Comment:
Woah, you can optionally call a function that might be null? 😮 cool!

Reply:
This is a TIL for me as well! I sometimes miss syntax sugar like this when working with Python.

Reply:
I think it was the Copilot's suggestion :)